File: tests/java/tests/java/org/elastos/hive/did/DIDEntity.java
package org.elastos.hive.did;

import org.elastos.did.DID;
import org.elastos.did.DIDDocument;
import org.elastos.did.DIDStore;
import org.elastos.did.RootIdentity;
import org.elastos.did.exception.DIDException;

import java.io.File;
import java.util.List;

class DIDEntity {
	private final String name;
	private final String phrasepass;
	protected final String storepass;

	private RootIdentity identity;
	private DIDStore store;
	private DID did;

	protected DIDEntity(String name, String mnemonic, String phrasepass, String storepass) throws DIDException {
		this.name = name;
		this.phrasepass = phrasepass;
		this.storepass = storepass;

		initPrivateIdentity(mnemonic);
		initDid();
	}

	protected void initPrivateIdentity(String mnemonic) throws DIDException {
		final String storePath = System.getProperty("user.dir") + File.separator + "data/didCache" + File.separator + name;

		store = DIDStore.open(storePath);

		String id = RootIdentity.getId(mnemonic, phrasepass);
		if (store.containsRootIdentity(id))
			return; // Already exists

		this.identity = RootIdentity.create(mnemonic, phrasepass, store, storepass);

		identity.synchronize(0);
	}

	protected void initDid() throws DIDException {
		List<DID> dids = store.listDids();
		if (dids.size() > 0) {
			this.did = dids.get(0);
			return;
		}

		DIDDocument doc = identity.newDid(storepass);
		this.did = doc.getSubject();
		System.out.format("[%s] My new DID created: %s%n", name, did);
	}

	protected DIDStore getDIDStore() {
		return store;
	}

	public DID getDid() {
		return did;
	}

	public DIDDocument getDocument() throws DIDException {
		return store.loadDid(did);
	}

	public String getName() {
		return name;
	}

	protected String getStorePassword() {
		return storepass;
	}

	@Override
	public String toString() {
		return this.did.toString();
	}
}

File: samples/ionic/src/app/hivejs/did/didentity.ts
import { RootIdentity, DIDStore, DID, DIDDocument } from "@elastosfoundation/did-js-sdk";
import { File, Logger } from "@elastosfoundation/hive-js-sdk";
import SdkContext from "../testdata";

export class DIDEntity {
    protected static LOG = new Logger("DIDEntity");

	private readonly name: string;
	private readonly phrasepass: string;
	private mnemonic: string
	protected storepass: string;

	private didStore: DIDStore;
	private did: DID;
	private didString: string;

	constructor (name: string, mnemonic: string, phrasepass: string, storepass: string, did?: string) {
		this.name = name;
		this.phrasepass = phrasepass;
		this.storepass = storepass;
		this.didString = did;
		this.mnemonic = mnemonic;
	}

	public async initDid(mnemonic: string, needResolve: boolean) {
		const path = SdkContext.RESOLVE_CACHE + File.SEPARATOR + this.name;
    console.log('DIDEntity.initDid: before store.open');
		this.didStore = await DIDStore.open(path);
    console.log('DIDEntity.initDid: before this.getRootIdentity');
		const rootIdentity = await this.getRootIdentity(mnemonic);
    console.log('DIDEntity.initDid: before this.initDidByRootIdentity');
		await this.initDidByRootIdentity(rootIdentity, needResolve);
	}

	protected async getRootIdentity(mnemonic: string): Promise<RootIdentity> {
		const id = RootIdentity.getIdFromMnemonic(mnemonic, this.phrasepass);
		return this.didStore.containsRootIdentity(id) ? await this.didStore.loadRootIdentity(id)
			: RootIdentity.createFromMnemonic(mnemonic, this.phrasepass, this.didStore, this.storepass);
	}

	protected async initDidByRootIdentity(rootIdentity: RootIdentity, needResolve: boolean): Promise<void> {
		const dids = await this.didStore.listDids();
		if (dids.length > 0) {
      console.log('DIDEntity.initDidByRootIdentity: got did');
			this.did = dids[0];
		} else {
			if (needResolve) {
        console.log('DIDEntity.initDidByRootIdentity: need resolve');
				const synced = await rootIdentity.synchronizeIndex(0);
				DIDEntity.LOG.info(`${this.name}: identity synchronized result: ${synced}`);
				this.did = rootIdentity.getDid(0);
			} else {
        console.log(`DIDEntity.initDidByRootIdentity: need create did, ${this.storepass}`);
				const doc: DIDDocument = await rootIdentity.newDid(this.storepass);
        console.log('DIDEntity.initDidByRootIdentity: before doc.getSubject');
				this.did = doc.getSubject();
				DIDEntity.LOG.info(`[${this.name}] My new DID created: ${this.did.toString()}`);
			}
		}
		if (!this.did) {
			DIDEntity.LOG.error("Can not get the did from the local store.");
		}
	}

	protected getDIDStore(): DIDStore {
		return this.didStore;
	}

	public getDid(): DID {
		return this.did;
	}

	public async getDocument(): Promise<DIDDocument> {
		return await this.didStore.loadDid(this.did);
	}

	public getName(): string {
		return this.name;
	}

	public getStorePassword(): string {
		return this.storepass;
	}

	public toString(): string {
		return this.did.toString();
	}
}
